### ISSUE TITLE
Wirecard-react-native question.

### ISSUE THREAD

**Issue by AliaksandrNazarov** [JoaoRSilverio/FastFoodFinderServer#3]
Hello. I didn't find another way to ask you about wirecard-react-native package. I want to use it in my RN app. And the main question is: when it will be production ready? At least approximately? Thanks.

**Comment by JoaoRSilverio** [comment]
Hi no problem , hope its something i can help with.

**Comment by AliaksandrNazarov** [comment]
And the main question is: when the 'wirecard-react-native' package will be production ready? At least approximately? Thanks.

**Comment by JoaoRSilverio** [comment]
Probably in a few months(1-2), right now its possible to make credit card
payments, but not other types of payments.

**Comment by AliaksandrNazarov** [comment]
Thanks for fast answer. Ok, I think our project needs just credit card payments, not more. Can I start using current rc2 version for implementation? Do you planning any breaking changes for prod version?

**Comment by JoaoRSilverio** [comment]
No, for prod version I plan having the other payment methods and the
project listed on GitHub. But no breaking changes.

**Comment by AliaksandrNazarov** [comment]
Thanks again, and have a nice day!

**Comment by JoaoRSilverio** [comment]
No prob, have a great day!

**Comment by AliaksandrNazarov** [comment]
Installed your package, did required steps. Builds failed on both platforms. Do i need any additional settings?

**Comment by AliaksandrNazarov** [comment]
Android:

**Comment by AliaksandrNazarov** [comment]
> Could not resolve all task dependencies for configuration ':app:debugRuntimeClasspath'.
   > Could not find com.wirecard.ecom:card:3.4.0.
     Searched in the following locations:
       - file:/Users/AliaksandrNazarau/.m2/repository/com/wirecard/ecom/card/3.4.0/card-3.4.0.pom
       - file:/Users/AliaksandrNazarau/.m2/repository/com/wirecard/ecom/card/3.4.0/card-3.4.0.jar
       - file:/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/react-native/android/com/wirecard/ecom/card/3.4.0/card-3.4.0.pom
       - file:/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/react-native/android/com/wirecard/ecom/card/3.4.0/card-3.4.0.jar
       - file:/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/jsc-android/dist/com/wirecard/ecom/card/3.4.0/card-3.4.0.pom
       - file:/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/jsc-android/dist/com/wirecard/ecom/card/3.4.0/card-3.4.0.jar
       - https://dl.google.com/dl/android/maven2/com/wirecard/ecom/card/3.4.0/card-3.4.0.pom
       - https://dl.google.com/dl/android/maven2/com/wirecard/ecom/card/3.4.0/card-3.4.0.jar
       - https://jcenter.bintray.com/com/wirecard/ecom/card/3.4.0/card-3.4.0.pom
       - https://jcenter.bintray.com/com/wirecard/ecom/card/3.4.0/card-3.4.0.jar
       - https://jitpack.io/com/wirecard/ecom/card/3.4.0/card-3.4.0.pom
       - https://jitpack.io/com/wirecard/ecom/card/3.4.0/card-3.4.0.jar
     Required by:
         project :app > project :wirecard-react-native

**Comment by AliaksandrNazarov** [comment]
Ios:
Nonnull' is disallowed with ARC
    if([paymentData objectForKey:@"orderID" != nil]){
                                 ~~~~~~~~~~~^~~~~~
/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/wirecard-react-native/ios/RNWirecard.m:174:34: warning: incompatible integer to pointer conversion sending 'int' to parameter of type 'id _Nonnull' [-Wint-conversion]
    if([paymentData objectForKey:@"orderID" != nil]){
                                 ^~~~~~~~~~~~~~~~~
In module 'Foundation' imported from /Users/AliaksandrNazarau/Documents/RB-mobileApp/ios/Pods/Headers/Public/React-Core/React/RCTDefines.h:9:
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.2.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSDictionary.h:17:46: note: passing argument to parameter 'aKey' here
- (nullable ObjectType)objectForKey:(KeyType)aKey;
                                             ^
/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/wirecard-react-native/ios/RNWirecard.m:177:48: error: implicit conversion of 'int' to 'id _Nonnull' is disallowed with ARC
    if([paymentData objectForKey:@"descriptor" != nil]) {
                                 ~~~~~~~~~~~~~~^~~~~~
/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/wirecard-react-native/ios/RNWirecard.m:177:34: warning: incompatible integer to pointer conversion sending 'int' to parameter of type 'id _Nonnull' [-Wint-conversion]
    if([paymentData objectForKey:@"descriptor" != nil]) {
                                 ^~~~~~~~~~~~~~~~~~~~
In module 'Foundation' imported from /Users/AliaksandrNazarau/Documents/RB-mobileApp/ios/Pods/Headers/Public/React-Core/React/RCTDefines.h:9:
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.2.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSDictionary.h:17:46: note: passing argument to parameter 'aKey' here
- (nullable ObjectType)objectForKey:(KeyType)aKey;
                                             ^
/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/wirecard-react-native/ios/RNWirecard.m:178:80: error: expected ';' after expression
        [order setDescriptor:(NSString * _Nullable) paymentData[@"descriptor"]]
                                                                               ^
                                                                               ;
/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/wirecard-react-native/ios/RNWirecard.m:180:21: error: no visible @interface for 'NSDictionary' declares the selector 'objectForKey:objectForKey:'
    if([paymentData objectForKey:@"orderID"!= nil || paymentData objectForKey:@"orderID"!= nil ]){
        ~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by JoaoRSilverio** [comment]
I'll look into this, can you you try using rc1 ? It seems to be related to
recent changes. The only dif between rc1 and rc2 is the orderId and
descriptor.

**Comment by AliaksandrNazarov** [comment]
Ok. I'll try rc1 version.

**Comment by AliaksandrNazarov** [comment]
Add new RNWirecardPackage() to the list returned by the getPackages() method

In MainActivity or MainApplication file?

**Comment by JoaoRSilverio** [comment]
In mainApplication.java in the android/app/src/main/java/com/yourapp folder

**Comment by AliaksandrNazarov** [comment]
Ok, I did as you wrote. 
Ios build is success with rc1
Android failed with same error

**Comment by JoaoRSilverio** [comment]
Try clearing android cache with "cd android && ./gradlew clean" from
react-native app folder

**Comment by AliaksandrNazarov** [comment]
I did it, and reset cache of bundler. Also reinstall node modules.

**Comment by AliaksandrNazarov** [comment]
Hello. Are you here?

**Comment by AliaksandrNazarov** [comment]
Trying to install package again.
Is manual installation required?

**Comment by AliaksandrNazarov** [comment]
Error without manual installation is same.
http://wirecard.github.io/paymentSDK-Android/integration/latest/Migration/
Read this docs, added:
maven { url  "https://dl.bintray.com/wirecard/ecom-android" } to build.gradle, and got new error:
Could not determine the dependencies of task ':wirecard-react-native:compileDebugAidl'.
> Could not resolve all task dependencies for configuration ':wirecard-react-native:debugCompileClasspath'.
   > Could not find com.facebook.react:react-native:0.59.10.
     Required by:
         project :wirecard-react-native

**Comment by AliaksandrNazarov** [comment]
As result for now:
1. added 
            mavenCentral()
            maven {
                url  "https://dl.bintray.com/wirecard/ecom-android"
            }
to:  my project foldes/android/build.gradle
2. Changed 
        implementation 'com.facebook.react:react-native:0.59.10'
    to 
        implementation 'com.facebook.react:react-native:+'
   in 
       node_modules/wirecard-react-native/android/build.gradle
After that build process goes deeper and failed in:
     node_modules/wirecardreactnative/android/src/main/java/com/rnwirecard/RNWirecardModule.java
Error:
Error: Command failed: ./gradlew app:installDebug -PreactNativeDevServerPort=8081
/Users/AliaksandrNazarau/Documents/RB-mobileApp/node_modules/wirecard-react-native/android/src/main/java/com/rnwirecard/RNWirecardModule.java:130: error: cannot find symbol
                .setCurrency(paymentInfo.getString("currency")).setDescriptor(paymentInfo.getString("descriptor"))
                                                               ^
  symbol:   method setDescriptor(String)
  location: class Builder
1 error

**Comment by AliaksandrNazarov** [comment]
Sometimes I hate RN))
I started the package on android, totally:
1. Added to android/build.gradle
   maven {
            url  "https://dl.bintray.com/wirecard/ecom-android"
        }
2. In node_modules/wirecard-react-native/android/build.gradle: 
implementation 'com.facebook.react:react-native:+'
3. In android/build.gradle:
    minSdkVersion = 21
    it was 16 by default, your package using 19. So, I need to set it to 19, but in the same time we need 
    to set 'multiDexEnabled true', 21's version has it by default
4. When i wrote about: 
   cannot find symbol
.setCurrency(paymentInfo.getString("currency")).setDescriptor(paymentInfo.getString("descriptor"))
^
symbol: method setDescriptor(String)

I using rc2 version, but rc1 hasn't this two methods: setDescriptor, setOrderId
If this methods are commented in rc2 version, the build is successful (but failed on ios)

Explain the points please. Thank you.

**Comment by JoaoRSilverio** [comment]
Hi , I added the correct steps to have a functional android version to the
readme also added some more working fields , will try to review the ios
instalation steps soon,and then upload a working example app on github.

**Comment by AliaksandrNazarov** [comment]
Hello. Good news! Thanks! I'll test rc3 now.

**Comment by AliaksandrNazarov** [comment]
Tested rc3, android is working after installation notes. Waiting updating for ios)
The ios build is failed now, with same error as in rc2 version.

**Comment by AliaksandrNazarov** [comment]
My refactored version of PaymentManager:
import ReactNativeWirecard, {
  IRNWirecardCcPayment,
  PAYMENT_METHODS,
  IOS_ENVIRONMENTS,
  ANDROID_ENVIRONMENTS,
  CURRENCYS,
  TRANSACTION_TYPES,
} from 'wirecard-react-native'
import { Platform } from 'react-native'

class PaymentManager {
  constructor() {
    const environment =
      Platform.OS === 'ios' ? IOS_ENVIRONMENTS.GERMANY_TEST : ANDROID_ENVIRONMENTS.GERMANY_TEST

    ReactNativeWirecard.initiateClient(this.onInitSuccess, this.onInitFail, environment)
  }

  onInitSuccess = () => {
    console.log('client was initiated successfully')
  }
  onInitFail = () => {
    console.log('client failed to initiate')
  }
  makePayment = () => {
    const mockPayment = {
      amount: '99',
      currency: CURRENCYS.EURO,
      // transactionType:TRANSACTION_TYPES.PURCHASE, // only purchase supported for now
      paymentMethod: PAYMENT_METHODS.CARD.paymentMethod,
      signature: 'baCkEndGeneRat3D5iGN4TuR3.==',
      requestID: 'my-internal-id',
      merchantID: 'wirecard-assigned-id',
      notificationUrl: 'https://my-backend.com/api/my-notifications-endpoint',
      setAttempt3d: false,
      setRecurring: false,
      setRequireManualCardBrandSelection: false,
      setAnimatedCardPayment: true,
    }
    // you can add a token to IRNWirecardCcPayment if you have it token: '2y3423423saa2';
    // this should trigger the Wirecard UI;
    ReactNativeWirecard.pay(
      () => this.onPaymentSuccess(),
      () => this.onPaymentFailure(),
      mockPayment
    )
  }

  onPaymentSuccess = (error, tokenId, transactionState, transactionId, requestId) => {
    console.log(`transaction${transactionId} done succesfully!`)
  }
  onPaymentFailed = (error, tokenId, transactionState, transactionId, requestId) => {
    console.log(`transaction failed :( ${error}`)
  }
}

export { PaymentManager }

And it working in this variant. I don't know why, but without wrapping into fn, this leads to an error:
 
![Снимок экрана 2020-06-10 в 12 51 26](https://user-images.githubusercontent.com/37554638/84254978-a3926080-ab1a-11ea-9075-af1456b5298b.png)

**Comment by AliaksandrNazarov** [comment]
I'm about this place:
ReactNativeWirecard.pay(
() => this.onPaymentSuccess(),
() => this.onPaymentFailure(),
mockPayment
)
And I removed bindings, we don't need it when using arrow fn.

**Comment by AliaksandrNazarov** [comment]
What part of wirecard docs I can use for testing? I'm about mockPayment data.
Now I'm trying to use info from here
https://doc.wirecard.com/CC_Main.html#CreditCard_PaymentModeCountriesandCurrencies
Right?

**Comment by JoaoRSilverio** [comment]
Yes, and then you can use these cards to test

https://doc.wirecard.com/CC_Main.html#API_CC_TestCards

**Comment by AliaksandrNazarov** [comment]
What did you use for cardholder name?

**Comment by AliaksandrNazarov** [comment]
And merchantId?

**Comment by AliaksandrNazarov** [comment]
For now I got only:
transaction failed :( General error occurred

And wirecard client never ask password.

**Comment by JoaoRSilverio** [comment]
It accepts any cardholder name
MerchantId:
33f6d473-3036-4ca5-8c64dac862d1

Also you have to generate the signature.
Theres a commented out method in the java file for this if you want to test.

**Comment by AliaksandrNazarov** [comment]
const mockPayment = {
      amount: '0.99',
      currency: CURRENCYS.EURO,
      // transactionType: TRANSACTION_TYPES.PURCHASE, // only purchase supported for now
      paymentMethod: PAYMENT_METHODS.CARD.paymentMethod,
      signature: 'baCkEndGeneRat3D5iGN4TuR3.==',
      requestID: 'testId',
      merchantID: '33f6d473-3036-4ca5-8c64dac862d1',
      notificationUrl: 'https://my-backend.com/api/my-notifications-endpoint',
      setAttempt3d: false,
      setRecurring: false,
      setRequireManualCardBrandSelection: false,
      setAnimatedCardPayment: true,
    }
![Снимок экрана 2020-06-10 в 16 11 19](https://user-images.githubusercontent.com/37554638/84271644-085ab480-ab35-11ea-92f3-89690deea1d6.png)

transaction failed :( General error occurred

**Comment by AliaksandrNazarov** [comment]
What I do wrong?
Signature should be a new every time?

**Comment by JoaoRSilverio** [comment]
You have to generate a signature the signature in the example wont work its
just a placeholder, in wirecard documentation it shows how to generate it.
Basically your backend should generate it in the production app, but for
testing theres a function in the wirecardModule.java that you can use. And
it will generate the signature in the java side.

**Comment by AliaksandrNazarov** [comment]
Thanks. My knowledge in java isn't deep or wide, can you explain to me the name of this function and how i can use it, please? (generate signature).

**Comment by AliaksandrNazarov** [comment]
RNWirecardModule - I search in this module, and can't understand what the function or method is.

**Comment by AliaksandrNazarov** [comment]
Can you share implementation of fn for generating signature, please?

**Comment by AliaksandrNazarov** [comment]
Hello. Did you test payment flow on android? Today my collaegue with me implemented signature method, tested it on data from example - got same result. But every time I got error: 
transaction failed :( General error occurred

**Comment by AliaksandrNazarov** [comment]
We successfully finished generating signature. The problem was in date format, we used format with milliseconds, but format should be without it.

**Comment by AliaksandrNazarov** [comment]
When you planning to fix ios builds?
And same question about customization UI.

**Comment by AliaksandrNazarov** [comment]
Thanks.

**Comment by JoaoRSilverio** [comment]
Yes I'm planning to look into it this weekend, and provide an react-native example app with working payments.
In regards to the customization I'm going to focus on testing and working examples first and then I'll add it.
